### PR TITLE
Allow multiple hostnames tags

### LINKED ISSUE #1570: forget should take multiple hostnames (since it can take multiple tags already)
The documentation at http://restic.readthedocs.io/en/latest/060_forget.html#removing-snapshots-according-to-a-policy explains how you can specify multiple tags when removing old backup revision. While you technically can specify multiple --hostname values as well, only the last is actually evaluated for the command.

## Output of `restic version`
```
restic 0.8.1
compiled with go1.9.2 on linux/amd64
```

## How did you run restic exactly?
```
backup my-remote forget --hostname hostname1 --hostname hostname2 --keep-hourly 4 --keep-daily 8 --keep-weekly 5 --keep-monthly 13 --keep-yearly 10
```

I am using https://github.com/binarybucks/restic-tools as a wrapper to easily backup to S3/minio.

## What backend/server/service did you use to store the repository?
S3

## Expected behavior
List snapshots to keep and delete for all hostnames.

## Actual behavior

It only lists snapshots from the last specified hostname.

## Steps to reproduce the behavior

see above

## Do you have any idea what may have caused this?

Looks like only the last hostname is evaluated for the forget command.

## Do you have an idea how to solve the issue?

Evaluate all hostnames (just like its already the case for tags).

## Did restic help you or made you happy in any way?

that's a very nice last point :-D. I've only been playing with restic for a brief time. But so far it looks very nice and extremely easy to setup. Looking forward to gather more experience with it.

### PR DESCRIPTION
**Pull request by aawsome**
What is the purpose of this change? What does it change?
--------------------------------------------------------

allows multiple hostnames for `dump`, `find`, `forget`, `ls`, `mount`,
`restore`, `snapshots`, `stats` and `tag`

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

closes #1570 

Checklist
---------

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [X] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: changelog/unreleased/issue-1570
Comment (rawtaz) [inline changelog/unreleased/issue-1570]: The title line can't end with a dot, it shouldn't have any punctuation. I also suggest shortening it, please consider changing these three lines to the following (note that I also changed `--hostname` to `--host` since the former is deprecated):

~~~
Enhancement: Support specifying multiple host flags for various commands

Previously commands didn't take more than one `--host` or `-H` argument into account, which could be limiting with e.g. the `forget` command.

The `dump`, `find`, `forget`, `ls`, `mount`, `restore`, `snapshots`, `stats` and `tag` commands will now take into account multiple `--host` and `-H` flags.
~~~

**Comment by rawtaz** [comment]
Nice work @aawsome! We need to adjust the changelog file a bit, once that's done I'll take a peek through the code again (just skimmed through it) and merge whenever possible.

File: internal/restic/snapshot_find.go
Comment (fd0) [inline internal/restic/snapshot_find.go]: Breaking up the complex condition makes the code so much easier to read, great idea!

File: internal/restic/snapshot_find.go
Comment (fd0) [inline internal/restic/snapshot_find.go]: Good name for the function! If `HasHostname()` has defined behavior for `hostnames` being an empty (or nil) slice the code here becomes even better :)

File: internal/restic/snapshot.go
Comment (fd0) [inline internal/restic/snapshot.go]: If `HasHostname()` accepts an empty slice and returns `true` in this case we can remove the `nil` checks below (needs to be mentioned in the description for the function though).

For slice emptyness, we should use `if len(hostnames) == 0`, this will match `hostnames == nil` as well as `hostnames == []string{}`.

File: cmd/restic/cmd_ls.go
Comment (fd0) [inline cmd/restic/cmd_ls.go]: Use `len(opts.Hosts) == 0` instead of checking equality to `nil`, that's more robust.

File: cmd/restic/find.go
Comment (fd0) [inline cmd/restic/find.go]: Use `len(hosts) == 0` here, too

File: internal/restic/snapshot_find.go
Comment (fd0) [inline internal/restic/snapshot_find.go]: Same here, the check for `nil` is not needed (and not entirely correct) as well.

**Review by fd0 — CHANGES_REQUESTED** [review]
Awesome, thanks for your contribution! There's a few things I added comments to, in addition it'd be great if you could:
 1. Squash the commits
 2. Use a proper email address for the commits instead of `alex@thinkpad` (use `git config --global user.email` or something similar)

Thanks again!

File: internal/restic/snapshot_find.go
Comment (aawsome) [reply internal/restic/snapshot_find.go]: I already thought about this and dismissed the idea for not being perfect consistent to what the name implies.. :smile: 
However, looking at `HasTagList()` there is already the same functionality. So I added this case, pointed it out in the description and also extended the description of `HasTagList()`. Now it is consistent that `Has"Object"` checks if one of the objects given in the list is satisfied and returns true for an empty list and `Has"Objects"` checks if all objects given in the list is satsified. IMO this makes sense.

File: internal/restic/snapshot.go
Comment (aawsome) [reply internal/restic/snapshot.go]: is changed, see above.

File: cmd/restic/cmd_ls.go
Comment (aawsome) [reply cmd/restic/cmd_ls.go]: is changed.

File: cmd/restic/find.go
Comment (aawsome) [reply cmd/restic/find.go]: is changed.

File: internal/restic/snapshot_find.go
Comment (aawsome) [reply internal/restic/snapshot_find.go]: is changed.

**Review by rawtaz — APPROVED** [review]
Approving this after @aawsome made changes requested by @fd0.

Edit: Apparently @fd0 has to approve it as well after having requested changes, I can't merge it as it is now. That's good!

**Review by fd0 — APPROVED** [review]

**Pull request merged**
